### ISSUE TITLE
transfers: Payment couldn't be completed (insufficient funds or no route to target)

### ISSUE THREAD

**Issue by drandreaskrueger** [raiden-network/raiden#1086]
Trying to transfer

* 1 BOKKYwei
* to: 0xfce399a738d8c9ab9817341213c5d411ea92caee
* BOKKY token contract 0x583cbbb8a8443b38abcc0c956bece47340ea1367

like this, no?
```
curl -X POST \
http://127.0.0.1:5001/api/1/transfers/0x583cbbb8a8443b38abcc0c956bece47340ea1367/0xfce399a738d8c9ab9817341213c5d411ea92caee \
-d '{"amount":1, "identifier":0}' --header 'Content-Type: application/json'
```
but the result is:
```
{"errors": "Payment couldn't be completed (insufficient funds or no route to target)."}
```

Balance is not the problem:

```
curl http://127.0.0.1:5001/api/1/address
{"our_address": "0xc62d9f5092737fa7689579e279a3a1dd5543fb9e"}
```
* [0.39999999999997667 BOKKY](https://ropsten.etherscan.io/token/0x583cbbb8a8443b38abcc0c956bece47340ea1367?a=0xc62d9f5092737fa7689579e279a3a1dd5543fb9e)
* [6.50449528 Ether](https://ropsten.etherscan.io/address/0xc62d9f5092737fa7689579e279a3a1dd5543fb9e)

So then it must be 'no route to target'. But:

```
curl -s http://127.0.0.1:5001/api/1/channels | jq
[
  {
    ...
  },
  {
    "partner_address": "0xfce399a738d8c9ab9817341213c5d411ea92caee",
    "balance": 21993,
    "reveal_timeout": 10,
    "settle_timeout": 600,
    "state": "opened",
    "channel_address": "0x177840481cd3063e9346169a819846600f493fee",
    "token_address": "0x583cbbb8a8443b38abcc0c956bece47340ea1367"
  }
]
```
and in the other direction (machine B to A):
```
curl -s http://127.0.0.1:5001/api/1/channels | jq
[
  {
    "partner_address": "0xc62d9f5092737fa7689579e279a3a1dd5543fb9e",
    "balance": 21993,
    "reveal_timeout": 10,
    "settle_timeout": 600,
    "state": "opened",
    "channel_address": "0x177840481cd3063e9346169a819846600f493fee",
    "token_address": "0x583cbbb8a8443b38abcc0c956bece47340ea1367"
  }
]

```

so there is a channel from `0xc62d9f5092737fa7689579e279a3a1dd5543fb9e` to `0xfce399a738d8c9ab9817341213c5d411ea92caee`, and it is funded. 

Why can't I transfer?  What am I not understanding yet?

Thanks.

### System description
same as [before](https://github.com/raiden-network/raiden/issues/1085)

**Comment by ChosunOne** [comment]
So I'm noticing that when I first start Raiden I can make one transfer, but upon attempting to do another I encounter the same bug.  

Logging [available_routes](https://github.com/raiden-network/raiden/blob/4493a19041c73959267518f055d078f3357f4528/raiden/raiden_service.py#L963) shows that the first time the route is there, but the second time it is empty.  It seems that the route is either not being stored or it is cleared once a transfer is done.  Subsequent attempts do not repopulate the route.

**Comment by hackaugusto** [comment]
> This appears to happen when the network_state changes from "unknown" to "unreachable". While the network state is "unknown" the transfer will succeed.

ATM we [allow transfers](https://github.com/raiden-network/raiden/blob/master/raiden/network/channelgraph.py#L183-L184) with nodes in the "unknown" network state, which is the [default value](https://github.com/raiden-network/raiden/blob/master/raiden/network/protocol.py#L310-L315) when the healthcheck is starting. This definitely explains what you experienced.

The network state is changed to "unreachable" after 20 seconds ([2 retries of 10 seconds each](https://github.com/raiden-network/raiden/blob/master/raiden/settings.py#L25-L26)). This may happen either because the ethereum node is [lagging behind](https://github.com/raiden-network/raiden/issues/1020#issuecomment-330422761) or the network is dropping packets.

Could you guys provide a bit of detail about your network setup? Thanks for the input.

**Comment by ChosunOne** [comment]
I'm currently running an ethereum node on my machine which is up to date to the rest of the blockchain.  The raiden node is on the same machine, and I am using the terminal to interact with both also on the same machine.  Other than that I've got about a dozen peers on my node and have gigabit internet connection.  Let me know if you need more specific information.

**Comment by hackaugusto** [comment]
> The raiden node is on the same machine, and I am using the terminal to interact with both also on the same machine.

Do you have a firewall that could be dropping UDP packets? We only send a 77 bytes Ping message and an 56 bytes ACK response from one node the other. The message is sent using the IP address given by UPNP or STUN published on the blockchain.

**Comment by ChosunOne** [comment]
hmm, is there a good way to check specifically for dropping udp packets with raiden?  using netstat -s reports that I've accepted many UDP packets and have no receive errors.

**Comment by konradkonrad** [comment]
Could you retest with explicit NAT traversal methods? I.e.
`--nat stun` and
`--nat upnp`

**Comment by ChosunOne** [comment]
The problem for me is persisting using both `--nat stun` and `nat --upnp`.  It seems to successfully create the NAT passthroughs for both, but when I attempt to transfer to the partner in the channel, I get the no route message.  However in my case I don't actually know if the partner node is online.  Is there a list of online nodes I can try opening a channel with aside from just joining the RTT network?

**Comment by konradkonrad** [comment]
> list of online nodes I can try opening a channel with 

you can [explicitly open channels with nodes](http://raiden-network.readthedocs.io/en/stable/api_walkthrough.html#opening-a-channel). for example try connecting directly with the echo node: `0x02f4b6bc65561a792836212ebc54434db0ab759a` ( http://raiden-network.readthedocs.io/en/stable/api_walkthrough.html#interacting-with-the-raiden-echo-node -- instead of joining the RTT-network, directly open a channel with the echo node)

**Comment by ChosunOne** [comment]
Ok, connecting to the echo node is reporting the following:
`curl -H "Content-Type: application/json" -X POST -d '{"amount":10}' http://localhost:5001/api/1/transfers/0x0f114a1e9db192502e7856309cc899952b3db1ed/0x02f4b6bc65561a792836212ebc54434db0ab759a`

`DEBUG:raiden.api.python initiating transfer amount=10 identifier=6749442843730964413 initiator=000e6dfc token=0f114a1e target=02f4b6bc
DEBUG:raiden.channel.netting_channel REGISTERED LOCK node=000e6dfc currentlocksroot=82646def lock_expiration=1894732 to=02f4b6bc lock_amount=10 lock_hashlock=05a21136 from_=000e6dfc lockhashes=['cb8e51cf', 'ebf46362', '85282fbe'] lockhash=847d09b2
DEBUG:raiden.channel.netting_channel REGISTERED TRANSFER node=000e6dfc nonce=4 transfer=<MediatedTransfer [channel:fd7c1984 nonce:4 transferred_amount:0 locksroot:d98c8ac0 hash:bd5d8fce id:6749442843730964413 hashlock:05a21136 expiration:1894732 amount:10]> to=02f4b6bc transferred_amount=0 from_=000e6dfc current_locksroot=d98c8ac0
DEBUG:raiden.network.protocol SENDING MESSAGE node=000e6dfc to=02f4b6bc message=<MediatedTransfer [channel:fd7c1984 nonce:4 transferred_amount:0 locksroot:d98c8ac0 hash:bd5d8fce id:6749442843730964413 hashlock:05a21136 expiration:1894732 amount:10]> echohash=5a379e38`  

But my curl request is hanging and not returning a 200 ok.

**Comment by drandreaskrueger** [comment]
> Could you retest with explicit NAT traversal methods? I.e.
> --nat stun and
> --nat upnp

Hooray - something is happening on the receiving end now.

But not looking good yet:
```
INFO:raiden.network.protocol	MESSAGE RECEIVED 
    node=fce399a7 
    message=<MediatedTransfer [channel:4f04666f nonce:10 transferred_amount:4 locksroot:3a29fa59 hash:b1bea3fa id:0 hashlock:4349a622 expiration:1984708 amount:1]> 
    message_sender=c62d9f50 
    echohash=3099c227

INFO:raiden.channel.netting_channel	INVALID NONCE 
     node=fce399a7 
     to=fce399a7 
     from_=c62d9f50 
     nonce=10
     expected_nonce=8

CRITICAL:raiden.network.protocol	maybe unwanted transfer 
    e=<MediatedTransfer [channel:4f04666f nonce:10 transferred_amount:4 locksroot:3a29fa59 hash:b1bea3fa id:0 hashlock:4349a622 expiration:1984708 amount:1]>
```

In the very beginning, I had a different answer:
```
INFO:raiden.network.protocol	MESSAGE RECEIVED 
    node=fce399a7 
    message=<Secret [channel:4f04666f nonce:9 transferred_amount:4 locksroot:00000000 hash:2f9d4c3d hashlock:60af6836]> 
    message_sender=c62d9f50 
    echohash=0c7553ef
```
that looked better - right? 

But I cannot replicate that anymore.


How can I reset the whole thing, and start over?


> Could you guys provide a bit of detail about your network setup? 

All our machines are behind the same external IP.

Is that a problem?

**Comment by drandreaskrueger** [comment]
How can I get rid of those 'maybe unwanted transfer' / 'INVALID NONCE' messages? They don't stop anymore.

**Comment by hackaugusto** [comment]
> How can I get rid of those 'maybe unwanted transfer' / 'INVALID NONCE' messages? They don't stop anymore.

This will happen regularly, it's only a problem if the nodes are not making any progress. If your nodes didn't make any progress and you *only* see invalid nonce messages, then please open a new issue.

#### Explanation

These log messages are shown because the same message is handled more than once. This happens because Raiden uses UDP protocol as a transport and it needs to resend the same message after a given timeout.

Under some circumstances the node may decide to send a packet a second time (or third, fourth ...) even though the previous packet was not dropped, only delayed, the end result is that the receiving node will process the same message more than once and it will complain. Again, this is perfectly fine as long as the nodes are making progress.

**Comment by hackaugusto** [comment]
> This appears to happen when the network_state changes from "unknown" to "unreachable". While the network state is "unknown" the transfer will seem to succeed, but the balance is not changed upon checking the channels via the api.

We don't send transfer to `unknown` nodes anymore (https://github.com/raiden-network/raiden/pull/1088). That doesn't fully solve the problem though.

Closing this issue in favor of https://github.com/raiden-network/raiden/issues/1133 to track health check problems.
